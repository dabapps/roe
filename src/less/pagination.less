
.pagination {

    .button {
        color: @pagination-button-color;
        background-color: @pagination-button-background;
    }

    .selected {
        color: @pagination-selected-color;
        background-color: @pagination-selected-background;
    }

    .next,
    .prev {
        color: @pagination-indicator-color;
        background-color: @pagination-indicator-background;
    }

    .dots {
        display: inline-block;
        height: auto;
        padding: @padding-small @padding-base;
        text-decoration: none;
    }

}
