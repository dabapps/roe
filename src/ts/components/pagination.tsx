import * as classNames from 'classnames';
import * as React from 'react';
import { PureComponent } from 'react';
import { ComponentProps } from '../types';
import Button from './forms/button';
import SpacedGroup from './spaced-group';

export interface PaginationProps extends ComponentProps {
  /**
   * className
   * @default
   */
  className?: string;
  /**
   * disabled
   * @default
   */
  disabled?: boolean;
  /**
   * pageSize
   * @default
   */
  pageSize: number;
  /**
   * currentPage
   * @default
   */
  currentPage: number;
  /**
   * itemCount
   * @default
   */
  itemCount: number;
  /**
   * changePage
   */
  changePage: (pageNumber: number) => void;
}

const LEFT_BUTTONS = 2;
const RIGHT_BUTTONS = 2;
const MAX_BUTTONS = LEFT_BUTTONS + RIGHT_BUTTONS + 1;

export class Pagination extends PureComponent<PaginationProps, {}> {
  public render() {
    const {
      className,
      disabled,
      itemCount,
      pageSize,
      currentPage,
      changePage,
      ...remainingProps
    } = this.props;

    return (
      <div {...remainingProps} className={classNames('pagination', className)}>
        <p className="display-inline">
          Showing {this.showingLowerCount()}-{this.showingUpperCounter()} of{' '}
          {itemCount}
        </p>

        <SpacedGroup className="float-right" large>
          <Button
            onClick={this.decrementPage}
            disabled={currentPage === 1 || disabled}
          >
            &#60;
          </Button>

          {this.paginationButtonCount().map((page: number, index: number) => {
            return (
              <Button
                key={index}
                className={this.getButtonType(page, index)}
                disabled={itemCount <= pageSize || disabled}
                onClick={this.onClickPageNumber(index, page)}
              >
                {this.getDisplayDots(index, page)
                  ? '...'
                  : this.getPageToGoTo(page, index)}
              </Button>
            );
          })}

          <Button
            disabled={
              !this.getMaxPages() ||
              currentPage === this.getMaxPages() ||
              disabled
            }
            onClick={this.incrementPage}
          >
            &#62;
          </Button>
        </SpacedGroup>
      </div>
    );
  }

  private decrementPage = () => {
    const { currentPage, changePage } = this.props;
    return changePage(currentPage - 1);
  };

  private incrementPage = () => {
    const { currentPage, changePage } = this.props;
    return changePage(currentPage + 1);
  };

  private showingLowerCount = () => {
    const { currentPage, pageSize } = this.props;
    return (currentPage - 1) * pageSize || 1;
  };

  private showingUpperCounter = () => {
    const { pageSize, currentPage, itemCount } = this.props;
    return pageSize * currentPage > itemCount
      ? itemCount
      : pageSize * currentPage;
  };

  private getButtonType = (page: number, index: number) => {
    const { currentPage } = this.props;

    if (currentPage === page) {
      return 'primary';
    }
    if (this.getDisplayDots(index, page)) {
      return 'blank';
    }

    return undefined;
  };

  private getMorePages = () => this.getMaxPages() > MAX_BUTTONS;

  private getMaxPages = () => {
    const { itemCount, pageSize } = this.props;

    return Math.ceil(itemCount / pageSize);
  };

  private getDisplayDots = (index: number, page: number) =>
    this.getMorePages() &&
    ((index === 1 && page > 2) ||
      (index === MAX_BUTTONS - 2 && page < this.getMaxPages() - 1));

  private getPageToGoTo = (page: number, index: number) => {
    if (this.getMaxPages() > MAX_BUTTONS && index === 0 && page > 1) {
      return 1;
    } else if (
      this.getMaxPages() > MAX_BUTTONS &&
      index === MAX_BUTTONS - 1 &&
      page < this.getMaxPages()
    ) {
      return this.getMaxPages();
    }

    return page;
  };

  private onClickPageNumber = (index: number, page: number) => {
    const { currentPage, changePage } = this.props;

    if (currentPage !== page && !this.getDisplayDots(index, page)) {
      return () => changePage(this.getPageToGoTo(page, index));
    }
  };

  private getStart = () => {
    const { currentPage } = this.props;

    return Math.max(
      Math.min(
        currentPage - LEFT_BUTTONS,
        this.getMaxPages() - LEFT_BUTTONS - RIGHT_BUTTONS
      ),
      1
    );
  };

  private getEnd = () =>
    Math.min(this.getStart() + MAX_BUTTONS, this.getMaxPages() + 1);

  private getRangeStaps = () => {
    const { itemCount, pageSize } = this.props;

    if (itemCount % pageSize === 0 && itemCount / pageSize < 5) {
      return Math.floor(itemCount / pageSize);
    }
    if (itemCount % pageSize !== 0 && itemCount / pageSize < 5) {
      return Math.floor(itemCount / pageSize) + 1;
    }

    return MAX_BUTTONS;
  };

  private paginationSeries = (start: number, end: number, steps: number) => {
    return Array.apply(null, { length: steps }).map(
      (item: number, index: number) =>
        Math.floor(start + index * ((end - start) / steps))
    );
  };

  private paginationButtonCount = () =>
    this.paginationSeries(this.getStart(), this.getEnd(), this.getRangeStaps());
}

export default Pagination;
