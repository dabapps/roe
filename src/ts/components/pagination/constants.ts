export const LEFT_BUTTONS = 2;
export const RIGHT_BUTTONS = 2;
export const MAX_BUTTONS = LEFT_BUTTONS + RIGHT_BUTTONS + 1;
