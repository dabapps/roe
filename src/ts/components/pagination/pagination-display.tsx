import * as classNames from 'classnames';
import * as React from 'react';

import { OptionalComponentPropAndHTMLAttributes } from '../../types';

export type PaginationDisplayProps = {
  /**
   * Number of items per page
   */
  pageSize: number;
  /**
   * Current page number to highlight (1 indexed)
   */
  currentPageNumber: number;
  /**
   * Total number of items available
   */
  itemCount: number;
} & OptionalComponentPropAndHTMLAttributes;

const PaginationDisplay = (props: PaginationDisplayProps) => {
  const {
    className,
    itemCount,
    pageSize,
    currentPageNumber,
    component: Component = 'p',
    ...remainingProps
  } = props;

  const lowerCount = (currentPageNumber - 1) * pageSize || 1;
  const upperCount =
    pageSize * currentPageNumber > itemCount
      ? itemCount
      : pageSize * currentPageNumber;

  return (
    <Component
      {...remainingProps}
      className={classNames('pagination-display', className)}
    >
      Showing {lowerCount}-{upperCount} of {itemCount}
    </Component>
  );
};

export default React.memo(PaginationDisplay);
