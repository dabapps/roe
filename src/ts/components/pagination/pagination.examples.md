#### Pagination example:

```js
class PaginationExample extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      page: 1,
    };

    this.changePage = this.changePage.bind(this);
  }

  changePage(value) {
    this.setState({ page: value });
  }

  render() {
    const ITEM_COUNT = 22;
    const PAGE_SIZE = 3;
    const { page } = this.state;

    return (
      <div>
        <Row>
          <Column xs={4}>
            <PaginationDisplay
              pageSize={PAGE_SIZE}
              currentPageNumber={page}
              itemCount={ITEM_COUNT}
            />
          </Column>
          <Column xs={8}>
            <Pagination
              className="float-right margin-top-base"
              pageSize={PAGE_SIZE}
              changePage={value => this.changePage(value)}
              currentPageNumber={page}
              itemCount={ITEM_COUNT}
            />
          </Column>
        </Row>
      </div>
    );
  }
}

<PaginationExample />;
```

#### Less variables

```less
@pagination-button-color: @grey-dark;
@pagination-button-background: @grey-lighter;
@pagination-selected-color: @white;
@pagination-selected-background: @primary;
@pagination-indicator-color: @grey-dark;
@pagination-indicator-background: @grey-lighter;
```
