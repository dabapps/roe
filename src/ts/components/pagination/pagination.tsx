import * as classNames from 'classnames';
import * as React from 'react';

import Button from '../forms/button';
import SpacedGroup from '../misc/spaced-group';
import {
  getPaginationSeries,
  getIsDots,
  getPageToGoTo,
  getButtonType,
} from './utils';

export type PaginationProps = {
  /**
   * Disable the pagination
   * @default false
   */
  disabled?: boolean;
  /**
   * Number of items per page
   */
  pageSize: number;
  /**
   * Current page number to highlight (1 indexed)
   */
  currentPageNumber: number;
  /**
   * Total number of items available
   */
  itemCount: number;
  /**
   * Next button text
   * @default '>'
   */
  nextText?: string;
  /**
   * Previous button text
   * @default '<'
   */
  prevText?: string;
  /**
   * Called when a page is selected
   */
  changePage: (pageNumber: number) => void;
} & React.HTMLAttributes<HTMLDivElement>;

interface PaginationButtonProps {
  totalPages: number;
  currentPageNumber: number;
  page: number;
  index: number;
  disabled: boolean | undefined;
  className: string | undefined;
  changePage: PaginationProps['changePage'];
}

const PaginationButton = ({
  currentPageNumber,
  totalPages,
  page,
  index,
  disabled,
  className,
  changePage,
}: PaginationButtonProps) => {
  const onClickPageNumber = React.useCallback(() => {
    /* istanbul ignore else */
    if (currentPageNumber !== page && !getIsDots(totalPages, index, page)) {
      changePage(getPageToGoTo(totalPages, page, index));
    }
  }, [totalPages, currentPageNumber, index, page, changePage]);

  return (
    <Button
      key={index}
      className={classNames(className, { disabled })}
      disabled={disabled}
      onClick={onClickPageNumber}
    >
      {getPageToGoTo(totalPages, page, index)}
    </Button>
  );
};

const PaginationButtonMemo = React.memo(PaginationButton);

const Pagination = (props: PaginationProps) => {
  const {
    className,
    disabled,
    itemCount,
    pageSize,
    currentPageNumber,
    changePage,
    nextText,
    prevText,
    ...remainingProps
  } = props;

  const pageCount = itemCount / pageSize;
  const totalPages = Math.ceil(pageCount);
  const isPrevButtonDisabled = currentPageNumber === 1 || disabled;
  const isNextButtonDisabled =
    !totalPages || currentPageNumber === totalPages || disabled;
  const isPageButtonDisabled = itemCount <= pageSize || disabled;

  const decrementPage = React.useCallback(() => {
    return changePage(currentPageNumber - 1);
  }, [currentPageNumber, changePage]);

  const incrementPage = React.useCallback(() => {
    return changePage(currentPageNumber + 1);
  }, [currentPageNumber, changePage]);

  return (
    <div {...remainingProps} className={classNames('pagination', className)}>
      <SpacedGroup className="pagination-group" large>
        <Button
          className={classNames('prev', {
            disabled: isPrevButtonDisabled,
          })}
          onClick={decrementPage}
          disabled={isPrevButtonDisabled}
        >
          {prevText ? <span className="prev-icon">{prevText}</span> : '<'}
        </Button>

        {getPaginationSeries(
          totalPages,
          pageCount,
          itemCount,
          pageSize,
          currentPageNumber
        ).map((page: number, index: number) => {
          const buttonType = getButtonType(
            totalPages,
            page,
            index,
            currentPageNumber
          );

          return getIsDots(totalPages, index, page) ? (
            <div key={index} className={buttonType}>
              ...
            </div>
          ) : (
            <PaginationButtonMemo
              key={index}
              index={index}
              page={page}
              totalPages={totalPages}
              currentPageNumber={currentPageNumber}
              className={buttonType}
              disabled={isPageButtonDisabled}
              changePage={changePage}
            />
          );
        })}

        <Button
          className={classNames('next', {
            disabled: isNextButtonDisabled,
          })}
          disabled={isNextButtonDisabled}
          onClick={incrementPage}
        >
          {nextText ? <span className="next-icon">{nextText}</span> : '>'}
        </Button>
      </SpacedGroup>
    </div>
  );
};

export default React.memo(Pagination);
