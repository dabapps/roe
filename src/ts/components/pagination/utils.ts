import { MAX_BUTTONS, LEFT_BUTTONS, RIGHT_BUTTONS } from './constants';

export const getIsDots = (
  totalPages: number,
  index: number,
  page: number
): boolean => {
  const hasMorePagesThanButtons = totalPages > MAX_BUTTONS;

  return (
    hasMorePagesThanButtons &&
    ((index === 1 && page > 2) ||
      (index === MAX_BUTTONS - 2 && page < totalPages - 1))
  );
};

export const getButtonType = (
  totalPages: number,
  page: number,
  index: number,
  currentPage: number
): string | undefined => {
  if (currentPage === page) {
    return 'selected';
  }
  if (getIsDots(totalPages, index, page)) {
    return 'dots';
  }

  return undefined;
};

export const getPageToGoTo = (
  totalPages: number,
  page: number,
  index: number
): number => {
  if (totalPages > MAX_BUTTONS && index === 0 && page > 1) {
    return 1;
  } else if (
    totalPages > MAX_BUTTONS &&
    index === MAX_BUTTONS - 1 &&
    page < totalPages
  ) {
    return totalPages;
  }

  return page;
};

const getStart = (totalPages: number, currentPage: number) => {
  return Math.max(
    Math.min(
      currentPage - LEFT_BUTTONS,
      totalPages - LEFT_BUTTONS - RIGHT_BUTTONS
    ),
    1
  );
};

const getEnd = (totalPages: number, start: number) =>
  Math.min(start + MAX_BUTTONS, totalPages + 1);

const getRange = (itemCount: number, pageSize: number, pageCount: number) => {
  const remainder = itemCount % pageSize;

  if (remainder === 0 && pageCount < MAX_BUTTONS) {
    return Math.floor(pageCount);
  }
  if (remainder !== 0 && pageCount < MAX_BUTTONS) {
    return Math.floor(pageCount) + 1;
  }

  return MAX_BUTTONS;
};

export const paginationSeries = (
  totalPages: number,
  pageCount: number,
  itemCount: number,
  pageSize: number,
  currentPage: number
): readonly number[] => {
  const start = getStart(totalPages, currentPage);
  const end = getEnd(totalPages, start);
  const range = getRange(itemCount, pageSize, pageCount);

  return [...Array(range)].map((_item, index) =>
    Math.floor(start + index * ((end - start) / range))
  );
};
