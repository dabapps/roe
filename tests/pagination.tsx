import * as enzyme from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Pagination } from '../src/ts/components/pagination/pagination';

describe('Pagination', () => {
  it('should match snapshot', () => {
    const instance = renderer.create(
      <Pagination
        className="float-right margin-top-base"
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={1}
        itemCount={26}
      />
    );

    expect(instance).toMatchSnapshot();
  });

  it('should disabled buttons', () => {
    const instance = renderer.create(
      <Pagination
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={1}
        itemCount={26}
        disabled
      />
    );

    expect(instance).toMatchSnapshot();
  });

  it('should render the first page link as an first option', () => {
    const instance = renderer.create(
      <Pagination
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={5}
        itemCount={70}
      />
    );

    expect(instance).toMatchSnapshot();
  });

  it('should display less than 5 buttons for minimal page counts', () => {
    const instance = renderer.create(
      <Pagination
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={1}
        itemCount={6}
      />
    );

    expect(instance).toMatchSnapshot();
  });

  it('should display next and previous text', () => {
    const instance = renderer.create(
      <Pagination
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={1}
        itemCount={6}
        prevText="Prev"
        nextText="Next"
      />
    );

    expect(instance).toMatchSnapshot();
  });

  it('should add extra button to cover remainder pages', () => {
    const instance = renderer.create(
      <Pagination
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={1}
        itemCount={7}
      />
    );

    expect(instance).toMatchSnapshot();
  });

  it('should increment pages on next click', () => {
    const instance = enzyme.shallow(
      <Pagination
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={1}
        itemCount={7}
      />
    );

    instance.find('.next').simulate('click');
  });

  it('should decrement pages on prev click', () => {
    const instance = enzyme.shallow(
      <Pagination
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={2}
        itemCount={7}
      />
    );

    instance.find('.prev').simulate('click');
  });

  it('should go the page by clicking the page number', () => {
    const instance = enzyme.mount(
      <Pagination
        pageSize={3}
        changePage={jest.fn()}
        currentPageNumber={2}
        itemCount={7}
      />
    );

    instance
      .find('.spaced-group.pagination-group')
      .childAt(1)
      .simulate('click');

    instance.unmount();
  });
});
