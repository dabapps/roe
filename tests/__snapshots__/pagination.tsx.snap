// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should add a "next" button when there are more than 3 pages 1`] = `
<div
  className="pagination"
>
  <span
    className="spaced-group large pagination-group"
  >
    <button
      className="button prev disabled"
      disabled={true}
      onClick={[Function]}
    >
      &lt;
    </button>
    <button
      className="button selected"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="button next"
      onClick={[Function]}
    >
      &gt;
    </button>
  </span>
</div>
`;

exports[`Pagination should disable buttons when disabled is true 1`] = `
<div
  className="pagination"
>
  <span
    className="spaced-group large pagination-group"
  >
    <button
      className="button prev disabled"
      disabled={true}
      onClick={[Function]}
    >
      &lt;
    </button>
    <button
      className="button selected disabled"
      disabled={true}
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="button disabled"
      disabled={true}
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="button disabled"
      disabled={true}
      onClick={[Function]}
    >
      3
    </button>
    <div
      className="dots"
    >
      ...
    </div>
    <button
      className="button disabled"
      disabled={true}
      onClick={[Function]}
    >
      9
    </button>
    <button
      className="button next disabled"
      disabled={true}
      onClick={[Function]}
    >
      &gt;
    </button>
  </span>
</div>
`;

exports[`Pagination should display custom next and previous text 1`] = `
<div
  className="pagination"
>
  <span
    className="spaced-group large pagination-group"
  >
    <button
      className="button prev disabled"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="prev-icon"
      >
        Prev
      </span>
    </button>
    <button
      className="button selected"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="button next"
      onClick={[Function]}
    >
      <span
        className="next-icon"
      >
        Next
      </span>
    </button>
  </span>
</div>
`;

exports[`Pagination should display less than 5 buttons for minimal page counts 1`] = `
<div
  className="pagination"
>
  <span
    className="spaced-group large pagination-group"
  >
    <button
      className="button prev disabled"
      disabled={true}
      onClick={[Function]}
    >
      &lt;
    </button>
    <button
      className="button selected"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="button next"
      onClick={[Function]}
    >
      &gt;
    </button>
  </span>
</div>
`;

exports[`Pagination should render page 1 as the first option, followed by an ellipse and further pages, when the current page is greater than 3 1`] = `
<div
  className="pagination"
>
  <span
    className="spaced-group large pagination-group"
  >
    <button
      className="button prev"
      onClick={[Function]}
    >
      &lt;
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      1
    </button>
    <div
      className="dots"
    >
      ...
    </div>
    <button
      className="button selected"
      onClick={[Function]}
    >
      5
    </button>
    <div
      className="dots"
    >
      ...
    </div>
    <button
      className="button"
      onClick={[Function]}
    >
      24
    </button>
    <button
      className="button next"
      onClick={[Function]}
    >
      &gt;
    </button>
  </span>
</div>
`;

exports[`Pagination should render the button for page 1 as selected when the currentPageNumber is 1 1`] = `
<div
  className="pagination float-right margin-top-base"
>
  <span
    className="spaced-group large pagination-group"
  >
    <button
      className="button prev disabled"
      disabled={true}
      onClick={[Function]}
    >
      &lt;
    </button>
    <button
      className="button selected"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      3
    </button>
    <div
      className="dots"
    >
      ...
    </div>
    <button
      className="button"
      onClick={[Function]}
    >
      9
    </button>
    <button
      className="button next"
      onClick={[Function]}
    >
      &gt;
    </button>
  </span>
</div>
`;
